### ISSUE TITLE
Tor and I2P integration

### ISSUE THREAD

**Issue by m3thm4th** [nicotine-plus/nicotine-plus#2491]
Would it be possible to integrate Tor and I2P tunneling in this application? For reference, [RetroShare](https://github.com/RetroShare) does this and it works very well.

**Comment by mathiascode** [comment]
It depends on what exactly you need. I'm not interested in maintaining any integration with Tor or I2P-related libraries, but if e.g. a bug in Nicotine+ prevents it from being used with a Tor proxy, I could look into fixing such issues.

**Comment by slook** [comment]
It would appear that RetroShare serializes encrypted packets of data in order to distribute them amoungst network peers...

https://github.com/RetroShare/RetroShare/blob/0bd4cdae19cdc56e0acde29ceeb7afef0132de14/retroshare-friendserver/src/friendserver.cc#L124-L154

... In this way it can receive, process and forward search result data as required without compromising the privacy of unrelated peers.

This approach would not be possible on slsk because its data is not encrypted nor signed with any sort of key, hence the reason why it is required for peers to connect directly with eachother in order to exchange search results and folder listings.

However, slsk does feature distributed search results and N+ has recently implemented this in PR #2097 see commit https://github.com/nicotine-plus/nicotine-plus/pull/2097/commits/5093cd1d52f1f8761b04478eae5a2df05d86f6e9 , it will be interesting to see how this develops once this is merged into stable releases.

@m3thm4th You might consider [testing latest master](https://github.com/nicotine-plus/nicotine-plus/blob/master/doc/TESTING.md) to investigate what happens to search results when you connect to peers with a client that is behind onion router(s). Take a look at [the "Connections" debug log](https://github.com/nicotine-plus/nicotine-plus/blob/master/doc/DEVELOPING.md#debug-logging) to analyze the network traffic patterns.

**Comment by DarkVoyage** [comment]
Just some remarks on using TOR with Soulseek.
1. TOR network doesn't endorse big file traffic. Same for torrents.
2. You will have a lot of connection problems, because many peers are sensitive to connection quality, ping and open ports.
3. You can use TOR in Soulseek through Proxifier, which can route traffic to SOCKS5 proxy.
4. Soulseek server and TOR network doesn't like each other for some reason, many exit nodes abort connection to server. You will have to configure a list of working exit nodes or find some other way to refresh circuit.

**Comment by slook** [comment]
Closing, since as discussed above there is no realistic possibility nor desire for such a complex integration.

**Comment by sjehuda** [comment]
https://github.com/nicotine-plus/nicotine-plus/issues/2491#issuecomment-2120356192

I agree that, Tor is unrelated, and not recommended. Also most sharers block Tor traffic.

In fact, however, I2P is, by design, intended to any kind of traffic, and multimedia in particular.

Presently, there are over five I2P based BitTorrent indexers, various of I2P based torrent trackers, and several P2P software, such as MuWire, qBitTorrent, Transmission-I2P, and XD.

By an analyses of three years, I confirm that, there are contents over BitTorrent (I2P), that are almost only accessible over I2P connectivity.

Of course, it is possible to distribute these over anywhere else, yet most of the sharers are only connected to I2P because it has a sufficient amount of data distributed.

https://i2pd.readthedocs.io/en/latest/tutorials/filesharing/

I firmly advise to add I2P, as it would certainly increase popularity and longer term viability.

**Comment by slook** [comment]
What specific feature(s) does Nicotine+ lack that prevents you from connecting it inside an I2P network?

**Comment by m3thm4th** [comment]
> What specific feature(s) does Nicotine+ lack that prevents you from connecting it inside an I2P network?

@slook Network proxy, to tunnel all connections

**Comment by slook** [comment]
From the limited research I have done about this, I don't think that providing a simple proxy option will work in the intended way...

https://i2p.net/en/docs/api/socks/

> _"Hoping that we can simply strap an arbitrary client on top of I2P without auditing both its behavior and its exposed protocols for security and anonymity is naive"..._

https://i2p.net/en/docs/overview/faq/#remote_i2cp

> _"if the program you intend to run sends this information as content, I2P has no way to protect your anonymity. For example, some mail applications will send the IP address of the machine"..._

Indeed, the unencrypted Soulseek protocol encapsulates addresses of peers within its headers which are exchanged in the clear. 

Without understanding exactly how this would work architecturally, it cannot be implemented easily using what we already have. Any implementation would seem to result in some degree of network fragmentation or other undesirable complexities.

It seems like it would only be effective if the application were completely redesigned for this purpose, or an entirely new network might end up having to be created...

https://i2p.net/en/docs/api/samv3/#version-3-protocol

We lack any proper understanding of it, unfortunately, so any potential integration would have to be approached by a different project.

Nevertheless, this is an interesting avenue of future research. Please do share any resources that you think may be relevant.

**Comment by m3thm4th** [comment]
Aside from Tor and I2P, the network proxy can be used with clearnet proxies to hide/change IP address, because it supports the classic proxy protocols (SOCKS4, SOCKS5, HTTP and HTTPS) so it's a very versatile feature that can also be used for anonymous networks, but each user can make of it anything they want. I renamed the issue.

**Comment by sjehuda** [comment]
I want to state this obvious statement.

SouldSeek clients, that are connected via I2P, can only communicate with SouldSeel clients and servers that are also connected via I2P.

I2P is a network of its own, and is not similar to proxy (e.g. Tor, VPN, et cetera)

Nevertheless, the title "Proxy integration" is best to realize this issue; and proxy support would certainly be useful; particularly to Nicotine+ and other SoulSeek software.

**Comment by sjehuda** [comment]
@mathiascode and @slook Shall we open this issue?

**Comment by sjehuda** [comment]
@slook

> It seems like it would only be effective if the application were completely redesigned for this purpose, or an entirely new network might end up having to be created...

> https://i2p.net/en/docs/api/samv3/#version-3-protocol

> We lack any proper understanding of it, unfortunately, so any potential integration would have to be approached by a different project.

Nicotine+ need not have to have anything more than integration of proxy.

Anything else is done by the I2P software (e.g. i2pd).

> Nevertheless, this is an interesting avenue of future research. Please do share any resources that you think may be relevant.

I advise, to support proxy integration, and test with a SoulSeek server which is operated over I2P.

It is significantly easier than you might think.

**Comment by slook** [comment]
> test with a SoulSeek server which is operated over I2P

Yes I would like to experiment with running a test Soulseek compatible network over I2P. We could all learn a lot from doing that.

Having been an I2P user for over a decade and from the research I have done about how it is maturing, that network looks like it is here to stay and so it represents a viable place if ever the network wanted or needed to migrate to another part of the internet for any reason.

> It is significantly easier than you might think.

I disagree. In this particular case it is significantly harder than you think because Soulseek is not like modern secure internetworking protocols that you might be familiar with. Also, while it may well be technically possible to do as you have suggested, we must ask ourselves if doing so is really desirable or could it perhaps lead to detrimental consequences for the health of the network. My main concerns are about the risks of peer fragmentation and excessive exposure of user data.

This would have to be done very carefully because Soulseek payloads are unencrypted and as such there is no inherent expectation of privacy when communicating with the server nor when peers make transmissions between themselves. This means the client cannot loosely attach itself onto a secure network via a simple proxy without risking its communications being broadcasted to uninvolved parties, so instead its core networking modules would have to be tightly integrated into whatever networks it participates in to allow safe unencrypted communication between only those who are explicitly involved with a particular transaction.

Rather than a simple proxy which might cause havoc unless users configured it in exactly the right way, in an ideal world the servers and clients themselves would all have to support more flexible network addressing schemes to know which route should be used to contact the servers and peers, such as what additional fields would be needed to support 128-bit IPv6 addressing could also be used to specify base 32 addresses and so on in addition to the currently supported 32-bit IPv4 addresses and ports, because this way it need not be necessary for all the users of any particular carrier network or addressing scheme to be isolated from one another.

**Comment by DarkVoyage** [comment]
What's the point of using a secure network in Soulseek, whose architecture presumes by design that you search for files through a centralized public server and your search is broadcasted to the entire network? It makes absolutely no sense. And if you add another network level, you lose access to all current users, which is the only reason for using Soulseek - to have access to existing userbase. The only way it can be made a de facto standard is if you make it an optional connection method, like uTP in torrents and maybe once upon a time users can be connectable using I2P. In fact I don't know much about it and don't see any use for myself.

**Comment by slook** [comment]
> What's the point of using a secure network in Soulseek...?

Good question.

The desire for a secure network comes mainly from an ever increasing number of users who must connect from behind restrictive local or carrier‑grade firewalls that they don’t control, which creates the “closed‑port” problem and prevents their client from acting as a file server. A VPN/VPS or other secure tunneling method, if configured correctly, lets them directly accept incoming peer connections through a bridge relay or reverse‑proxy on a remote network, rather than having to rely on the server to negotiate indirect connection handshaking for them.

Even though the server address is editable in the client, it is mainly useful only for development and testing of [Soulfind](https://github.com/soulfind-dev/soulfind), where developers manually enter a different hostname or IP address, such as:
<img width="689" height="47" alt="Image" src="https://github.com/user-attachments/assets/1ea484be-dbc6-4e17-acfe-cbb3e1e00d39" />

Some users can apparently evade DNS blocks by entering the server’s IPv4 address directly, though the official server lacks an IPv6 address and the current protocol does not support IPv6 peers. Nevertheless, IPv6 support is planned in Nicotine+ for the sake of testing [soulfind#60](https://github.com/soulfind-dev/soulfind/issues/60) and we hope it will eventually be adopted upstream.

Adding proper IPv6 support would require a protocol extension with new fields (i.e. 128‑bit address fields and adjacent spare padding fields with null terminators) to accommodate encoding of future secure address types such as base‑32 (like on [I2P](https://i2p.net)) or other SHA hashes (like on [RNS](https://reticulum.network/)) without causing errors in older clients, so we shall always need to support IPv4 fallback.

Meanwhile, the OP desires an additional entry to simplify the implementation of a proxy server which is intended to workaround some kind of connectivity problems in their network environment. However, since the logic is not well understood then I wonder if this would make the client too open to misconfiguration or misuse.

This is not motivated by a need for extra security on the existing network, but supporting secure network addresses could help the client operate over, under, through or around ISP restrictions by tunneling connections through a carrier network that embeds security directly into its addressing scheme.

> if you add another network level, you lose access to all current users

Indeed.

The traditional unencrypted nature of the Soulseek protocol is a fundamental part of it beauty and elegance, since this attribute makes it entirely unsuitable to use for transmitting illicit materials or contraband so it therefore requires no moderation at the network operator level.

Most users agree that it works pretty well as it is, so there are certainly no plans to change these characteristics nor any intention of making an alternative network layer separated from the main network.

> In fact I don't know much about it and don't see any use for myself.

Whilst we may not need to use the proposed proxy features ourselves, we must consider the reasons why there does seem to be a large number of users wishing to establish themselves on the network by using highly elaborate connection routing topologies, and what Nicotine+ might be able to do to help them overcome the problems that they are faced with.

They are saying that the client is severely throttled or totally inoperative in their network environment because something basic is missing from the program. However it is more likely that the overall design of the protocol is unsuitable for their special use case.

As such there is not really a desire to make the client-server nor peer-to-peer connections themselves secure, maybe we are talking about integrating modern user-friendly routing options to make a secure connection between a user's terminal and their client.

A good example of a working implementation of this is the [Relay Mode](https://github.com/slskd/slskd/blob/master/docs/relay.md#what-is-relay-mode) of `slskd`, but sadly that is not user-friendly, so the N+ project mission is to make innovations which are accessible for ordinary users who don't have a degree in network engineering.

**Comment by DarkVoyage** [comment]
Thanks for detailed explanation. Closed port is a problem for all P2P, because two peers with closed ports are unconnectable to each other. I found solution in VPN with port forwarding. They rarely allow opening ports, but some do.

IPv6 looks to me a failed project. It grows for almost 20 years, but I don't see any point, when it replaces IPv4. Whole internet is still mostly IPv4.

**Comment by sjehuda** [comment]
@slook

I want to remark, that all software must be over I2P (client and server), as you probably already know.

Also, when I stated, that it be easier than we might think, I mean that the prime task is to connect to "proxy" (I2P), and from that stage, we can test, inspect, and explore the activity of SoulSeek over I2P.

**Comment by slook** [comment]
> IPv6 looks to me a failed project. It grows for almost 20 years, but I don't see any point

You can get a cheap VPS (upon which you can make your own truly private VPN) that comes with several free IPv6 addresses, whereas a static IPv4 address always costs quite a lot extra.

> Whole internet is still mostly IPv4.

Nowadays most things are dual stack, because indeed there are still too many end points that cannot work on IPv6.

When IP was first implemented by ARPA address numbers were only 8-bit as it was originally thought that there would not ever be more than 256 networked computer systems in the world. The first 32 numbers were used and rest were reserved to allow for future expansion. I guess it wasn't really planned out very well!

> I want to remark, that all software must be over I2P (client and server), as you probably already know.

Yes indeed. That would be what would happen if we tried it with the software that is available now.

But nothing prevents a server from being a bridge to translate traffic between several totally different networks, and if compatible peers are optionally connected together with each other over I2P then I don't see why this should have to isolate them from the server where everyone else is, because a client could be made to link different types of connections at the same time if the machine has multiple network interfaces.

**Comment by sjehuda** [comment]
> But nothing prevents a server from being a bridge to translate traffic between several totally different networks.

Indeed. Yet, it would cost significant resources to the operator of such (proxy) server.

**Comment by slook** [comment]
Not really, because all the clients are effectively little servers in their own right so they could negotiate all the various types of peer connections directly by themselves without the server being needed to act as a fully-fledged proxy in the traditional sense.

In any case, no actual user generated content is ever processed by the server (the network cannot be designed like that for legal reasons) so it is only serving the lightweight tasks of user authentication, plaintext chat relay and a directory of peers (and perhaps in future, proxying a directory of directories).

Several different connection options could potentially be offered by a server according to what kind of network as to the preference or capabilities of each client, so in that case there would not be any more than the usual number of client connections to it (and if some clients want to have simultaneous sockets then only one of them - ideally the fastest one - would actually be used at a time).

Furthermore, there could be many load balanced servers with some kind of secure synchronization channel between them (any protocol could be used for server-server intercoms because that implementation detail doesn't really matter to the end clients) then the resources could actually be very low by modern standards, and even by legacy standards (considering the original server software and protocol was developed and deployed for several years on a Pentium 4 desktop server).

For example the compiled [Soulfind](https://github.com/soulfind-dev/soulfind/releases/tag/latest) binary is less than 1MB and it can incrementally vacuum its database whilst running (max size allow, say, 1 or 2 GB of bandwidth per day for up to 100000 users = ~50GB per month) and happily runs on any single-core machine such as an old smartphone (running Termux) or a Raspberry Pi or low-end VPS (running a [barebones Linux or BSD distro](https://tierhive.com/blog/tierhive-howto/how-to-run-alpine-with-just-23mb-ram) - the only software dependency is [SQLite](https://sqlite.org/) that is about 5MB). So I think you could do it comfortably in well under 2GB RAM and 5GB SSD plus 10GB HDD for backups. Nothing crazy is needed for the CPU (a low-end Xeon or ARM would do just fine) since it is only single-core software and obviously there is no use for a GPU whatsoever.

I don't know if the highest load is caused by processing login attempts, search queries or chatting, but either way we are only talking about small UTF-8 strings of up to a certain maximum size (I think its limited up to 4096 bytes per message but usually most queries are much less than 1024 bytes). The most expensive operation is the password hashing, but that is only done once per session.

So you could run several instances of those on a low-end VPS for less than about $5 per month in total for all of them, or you could do it for nothing if you already have access to several decent residential or commercial broadband connections. You could even put some on dialup or over packet radio to cover remote areas that don't have the internet. If you powered your servers from batteries and charged them with solar then that could make the network available in off-grid locations without reliable electricity (idk, like festivals or techno raves in war zones and whatnot) whilst reducing the running costs down to only the cost of your equipment.

Once developed and the reliability is proven, then I suppose the logical progression from running server(s) is to just build the distributed server software right into the clients so all the capable peers (i.e. those with open ports) can act as load balancers via some kind of elaborate global routing arrangement. I think that kind of network topology is called [federation](https://en.wikipedia.org/wiki/Federation_(information_technology)) _Edit: I just learned a new-skool phrase [Fediverse](https://en.wikipedia.org/wiki/Fediverse)_. Then the ongoing costs of running such a network would be zero (obviously not counting the many hours of software developer's time, of course).

Anyway, whilst it is interesting to ponder about the future of server-side operations, it is off-topic to this issue about enabling users to more easily proxy the connection that goes between their client and their terminal.
